qemu.go

```diff
@@ -985,6 +996,110 @@ func (q *qemu) hotplugRemoveDevice(devInfo interface{}, devType deviceType) erro
 	return q.pod.storage.storeHypervisorState(q.pod.id, q.state)
 }
 
+func (q *qemu) hotplugCPU(amount uint32, op operation) error {
+	defer func(qemu *qemu) {
+		if q.qmpMonitorCh.qmp != nil {
+			q.qmpMonitorCh.qmp.Shutdown()
+		}
+	}(q)
+
+	qmp, err := q.qmpSetup()
+	if err != nil {
+		return err
+	}
+
+	q.qmpMonitorCh.qmp = qmp
+
+	// Add CPUs
+	if op == addDevice {
+		// let's see the list of hotpluggable CPUs
+		hotpluggableCPUs, err := q.qmpMonitorCh.qmp.ExecuteQueryHotpluggableCPUs(q.qmpMonitorCh.ctx)
+		if err != nil {
+			return fmt.Errorf("failed to hotplug CPUs: %v", err)
+		}
+
+		var hotpluggedCPUs uint32
+		for _, hc := range hotpluggableCPUs {
+			// qom-path is the path to the CPU, non-empty means that this CPU is already in use
+			if hc.QOMPath != "" {
+				continue
+			}
+
+			driver := hc.Type
+			cpuID := fmt.Sprintf("cpu-%d", len(q.state.HotpluggedCPUs))
+			socketID := fmt.Sprintf("%d", hc.Properties.Socket)
+			coreID := fmt.Sprintf("%d", hc.Properties.Core)
+			threadID := fmt.Sprintf("%d", hc.Properties.Thread)
+			if err := q.qmpMonitorCh.qmp.ExecuteCPUDeviceAdd(q.qmpMonitorCh.ctx, driver, cpuID, socketID, coreID, threadID); err != nil {
+				// don't fail, let's try with other CPU
+				continue
+			}
+
+			q.state.HotpluggedCPUs = append(q.state.HotpluggedCPUs, CPUDevice{cpuID})
+			hotpluggedCPUs++
+			if hotpluggedCPUs == amount {
+				// All CPUs were hotplugged
+				return nil
+			}
+		}
+
+		return fmt.Errorf("failed to hotplug CPUs: only %d CPUs were hotplugged", hotpluggedCPUs)
+	}
+
+	// Remove CPUs
+	if amount > uint32(len(q.state.HotpluggedCPUs)) {
+		return fmt.Errorf("Unable to hotunplug %d CPUs, only %d CPUs can be hotunplugged", amount, len(q.state.HotpluggedCPUs))
+	}
+
+	var hotunpluggedCPUs uint32
+	var hotpluggedCPUs []CPUDevice
+	copy(q.state.HotpluggedCPUs, hotpluggedCPUs)
+	for i, huc := range hotpluggedCPUs {
+		if err := q.qmpMonitorCh.qmp.ExecuteDeviceDel(q.qmpMonitorCh.ctx, huc.ID); err != nil {
```

Comment:
Again which one are you deleting. What is the mapping between this and what got offlined in the guest kernel. How are you co-relating it.